Issued Asset Fees

tomt1664 opened the pull request:
This ELIP describes proposed features for Elements that enable the use of issued assets for transaction fees in addition to the policy asset (pegged asset). 

This will enable an Elements blockchain/sidechain to use any issued asset for the transaction fees, with the assets used and rates applied being agreed and set by the block creators. Changes are proposed to relay/mempool policy and block creation. 

Several mechanisms are proposed to set, synchronise and publish the assets and rates that are accepted by the block producers.

tomt1664 committed:
Create ELIP 202 for issued asset fees

Add controller descritpion

Add publication descritpion

Add test vectors

fix typos

Edit README for elip 0202

Fix path

fixed errors and renamed rate to multiplier

updated fee rate multiplier logic and limits

updated example, removed repetitions, removed elip number

change filename temporarily

Add mempool ordering logic for changing rate multipliers.

updated abstract and motivation sections

modifified the controler update specification section to clarify the nature of the controller transaction

removed specification of specific config options and RPCs

removed implementation details not relevant to spec

added statment about rounding errors as multiplier approaches minimum

changed syntax for the multiplier and rate arithmetic to rust

changed references to single fee outputs to general fees and added statement on handling mutiple fee assets in a single transaction

statement added concerning rate publication and ibd

added clarification about separation of controller and publication features

added rules for the application of conflicting controller updates

update of publication statement

Co-authored-by: Byron Hambly <byron@hambly.dev>

add units to effective_fee_rate

Co-authored-by: Byron Hambly <byron@hambly.dev>

fix typo

Co-authored-by: Byron Hambly <byron@hambly.dev>

specify big endian for encoded fee rate multiplier

updated ELIP number to 204

delta1 reviewed (commented):
First pass, a few minor things

apoelstra commented:
In 5f0dbbd6ca01ae824d2e779383884768b62e3cf6:

trailing whitespace on most lines

apoelstra commented:
In 9641693b88a3bb0c36e8a602c6ba84a8b625594a:

This introduces the terms "input" and "output" without defining them, and still says nothing about how overflow in feerates is handled.

>The minRelay rate or min mempool rate would have be bigger than ASSET_RATE_SCALE_FACTOR. But a check for this should be added.

It would just need to be greater than one. But my question was not about min relay, but about how these transactions' feerates are computed so that transactions can be ordered in the mempool (AFAICT this is never mentioned in this document).

tomt1664 commented:
Sorry for the delay in replying. I've gone through and changed the logic of how the rate multiplier is applied, changing it to computing an 'effective rate' for a transaction with an issued asset fee which can then be used in place of the policyAsset rate for both meeting relay/mempool minimums and mempool ordering. 

> This introduces the terms "input" and "output" without defining them, and still says nothing about how overflow in feerates is handled.

I've removed that, and specified clearly the format of values used in configuration and RPC arguments and how it is then converted to the fixed precision integer. 

>and still says nothing about how overflow in feerates is handled. 

> It would just need to be greater than one. 

Yes, you are right. I've added a maximum value parameter for the rate multiplier. 
In the calculation for the new effective fee rate, the rate multiplier is multiplied by the fee rate in the issued asset. This could potentially overflow int64, but only if the fee rate in the issued asset multiplied by the multiplier gives an effective fee rate of > 922 * COIN (a value of 922 LBTC/vbyte). It's implausible a rate multiplier and fee would be set to this level on purpose, but to protect against that possibility, if the result of the multiplication does exceed that value, then the effective fee rate is set to this maximum. 

> But my question was not about min relay, but about how these transactions' feerates are computed so that transactions can be ordered in the mempool (AFAICT this is never mentioned in this document). 

The 'effective rate' for an issued asset fee transaction is now used for both comparison with minimums and also for mempool ordering. Statements explaining this have been added to the text.

apoelstra commented on elip-0202.mediawiki:
```diff
@@ -0,0 +1,232 @@
+
+For fees outputs in an issued asset, the fee rate (in units/vbyte) will be scaled by a fee rate multiplier to determine an 'effective rate'. This effective rate is then used in place of the policyAsset fee rate for the purpose of meeting relay and mempool minimums and for mempool ordering. 
+
+E.g. An issued asset is USDT. The current conversion rate between USDT and LBTC (<code>policyAsset</code>) is 100:1 (i.e. 100 USDT to one LBTC) the issued asset fee rate multiplier would then be set to 0.01 (i.e. the fee paid in USDT will be multiplied by 0.01 for comparison with LBTC fees). If the USDT asset is used for the fee in any transaction, it should pay 100 times in USDT base units the f ...
```
In 4b5eeb206931975787cb5d67114ecd4ed8d1557c:

I don't undersztand this example. For one thing, it seems cherry-picked to have an integer ratio; what if the actual exchange rate is `105:100` say? For another thing, it seems to normalize the fee to USDT rather than L-BTC. If there are multiple fee outputs with different assets then which one is chosen to normalise to?

tomt1664 replied:
A simple integer ratio was chosen to make the arithmetic of this introductory example easier, but maybe it's too abstract. A more realistic exchange rate can be used, which illustrates precision/rounding issues. 

The fee is paid in USDT, and the monetary value of that fee must equal (or exceed) the monetary value of the fee currently required in LBTC. The multiplier converts the monetary value of an amount of USDT to an amount of LBTC. 

Another attempt:

```
E.g. An issued asset is USDT. The current accepted exchange rate between USDT and LBTC (the <code>policyAsset</code>) is $114,171.23 to 1 LBTC. The issued asset fee rate multiplier for this ratio is then set to <code>1/114,171.23 = 0.00000876</code> (rounded to a precision of 1e-8). For a transaction that uses the USDT asset for the fee, the fee rate (in units/vbyte) multiplied by this number must be greater than or equal to the minimum fee rate (in sat/vbyte) applied in LBTC. 

If the current minimum relay fee rate is 1 sat/vbyte in LBTC, then the minimum fee rate required in USDT would be <code>1/0.00000876 = 114156</code> units/vbyte (rounded up to the nearest integer). To relay this transaction, a client will determine the fee rate in the USDT asset, multiply it by 0.00000876 to calculate the effective rate and then verify it is greater than or equal to 1 sat/vbyte. 
```

In the case of multiple fee outputs, in principle, an effective fee (the fee amount scaled by the multiplier for each different asset) can be calculated for each fee output and then an effective fee rate for the transaction determined. However, it may be simpler to require wallets to use a single asset and output, and only consider the first one.

apoelstra commented on elip-0202.mediawiki:
```diff
@@ -0,0 +1,232 @@
+
+A new configuration option will be added to enable any node that is creating blocks to publish the assets, rates (and optionally destination) they are applying to their own policy. This information will be encoded into an <code>OP_RETURN</code> output for each accepted asset and included in the coinbase transaction for blocks that they create every <code>N</code> blocks (where <code>N</code> is s ...
+
+In order for the acceptance of issued asset and rates for fees to be reliable and predictable, all block creators and bridging nodes (and all relaying nodes) should be applying the same policy with the same list of accepted assets and rates. To coordinate this and apply changes to a large number of nodes is impractical, which may be required regularly as exchange rates change. To provide an alter ...
```
In 4b5eeb206931975787cb5d67114ecd4ed8d1557c:

This paragraph doesn't make sense after the previous paragraph. It says that the blocksigners will encode exchange rates in `OP_RETURN` outputs from coinbases, then describes an "alternate" method in which exchange rates are accepted "via an on-chain transaction" with no further details given. Is this the same thing as the OP_RETURN mechanism you just described? It is contrasted to "requiring direct access to the RPC interface" but you've only described using the RPC interface to access the current rates, not for setting them.

tomt1664 replied:
> This paragraph doesn't make sense after the previous paragraph. It says that the blocksigners will encode exchange rates in `OP_RETURN` outputs from coinbases, then describes an "alternate" method in which exchange rates are accepted "via an on-chain transaction" with no further details given. Is this the same thing as the OP_RETURN mechanism you just described?

These are two separate features. The first paragraph introduces a method for publishing the assets and rates accepted by a node. The second is a separate method for remotely *setting* the assets and rates accepted by a node without direct access to it, as an alternative to configuring node directly. 

> It is contrasted to "requiring direct access to the RPC interface" but you've only described using the RPC interface to access the current rates, not for setting them.

It states above "This can be set as part of the node configuration, specifying the <code>assetID</code>, the relative rate multiplier (and optionally the coinbase destination for the fee collection by the block creator) before initialization, or set dynamically at runtime via a new RPC option."

delta1 reviewed (approved):
preliminary utACK bcd0cb8620e8e98dfded56342fc5e523bdbd1a63; some minor comments/suggestions

psgreco commented:
Carrying over @delta1 's ack after the nits/rebase
